sycl/doc/SYCLPluginInterface.md

```diff
@@ -0,0 +1,178 @@
+# The SYCL Runtime Plugin Interface.
+
+
+## Overview
+The SYCL Runtime Plugin Interface (PI) is the interface layer between
+device-agnostic part of the SYCL runtime and the device-specific runtime layers
+which control execution on the device. It employs the “plugin” mechanism to bind
+to the device specific runtime layers similarly to what is used by libomptarget
+or OpenCL.
+
+![PI in SYCL runtime architecture](images/SYCL_RT_arch.svg)
+
+The plugin interface and the discovery process behind it allows to dynamically
+plug in implementations based on OpenCL and “native” runtime for a particular
+device – such as OpenCL for
+FPGA devices or native runtimes for GPUs. Implementations of the PI are
+“plugins” - dynamic libraries or shared objects which expose a number of entry
+points implementing the PI interface. The SYCL runtime collects those function
+pointers into a PI interface dispatch table - one per plugin - and uses this
+table to dispatch to the device(s) covered by the corresponding plugin. Further
+sections give complete descriptions of all entry points as well as the dispatch
+table.
+
+PI is based on a subset of OpenCL 1.2 runtime specification, it follows its
+platform, execution and memory models in all aspects except those explicitly
+mentioned in this document. A part of PI API types and functions have exact
+matches in OpenCL. Whenever there is such a match, the semantics also fully
+matches unless the differences are explicitly specified in this document. While
+PI has roots in OpenCL, it does have many differences, and the gap is likely
+to grow, for example in the areas of memory model and management, program
+management.
+
+## Discovery and linkage of PI implementations
+
+![PI implementation discovery](images/SYCL_plugin_discovery.svg)
+
+Device discovery phase enumerates all available devices and their features by
+querying underlying plugins found in the system. This process is only performed
+once before any actual offload is attempted.
+
+### Plugin discovery
+
+Plugins are physically dynamic libraries stored somewhere in the system where
+the SYCL runtime runs. TBD - describe the process in details.
+
+#### Plugin binary interface
+TBD - list and describe all the symbols plugin must export in order to be picked
+up by the SYCL runtime for offload.
+
+#### OpenCL plugin
+
+OpenCL plugin is a usual plugin from SYCL runtime standpoint, but its loading
+and initialization involves a nested discovery process which finds out available
+OpenCL implementations. They can be installed either in the standard Khronos
+ICD-compatible way or not, and the OpenCL plugin can hook up with both.
+
```

Comment:
It sounds that the "OpenCL plugin" would support ANY (not "both") OpenCL implementation, no matter how that is installed, which can't be true. Maybe elaborate on what exactly you meant, or just add TBD comment.

Reply:
Oh, I see you describe "both" ways below, which sounds fine for now.